caffe2 - support flaky operator tests for caffe2 build

duc0 opened the pull request:
Summary:
- Make a python decorator caffe2_flaky for caffe2 opreator unit tests.
- The environment variable CAFFE2_RUN_FLAKY_TESTS are now used to mark flaky test mode

During test run,
- If flaky tests mode are on, only flaky tests are run
- If flaky tests mode are off, only non-flaky tests are run

Mark ctc_beam_search_decoder_op_test as flaky

Reviewed By: salexspb

Differential Revision: D14468816

ezyang commented on caffe2/python/test_util.py:
```diff
@@ -60,6 +77,15 @@ def setUpClass(cls):
+        if (is_flaky_test_mode() and not is_flaky_test):
+            raise unittest.SkipTest("Non-flaky tests are skipped in flaky test mode")
+        elif (not is_flaky_test_mode() and is_flaky_test):
+            raise unittest.SkipTest("Flaky tests are skipped in regular test mode")
```
FYI: after I added this to PyTorch's top-level TestCase class, I ended up having to fix a lot of subclasses to actually call `setUp` on the superclass (they overrode without calling super). You should audit this in Caffe2 as well.

duc0 committed:
caffe2 - support flaky operator tests for caffe2 build (#18155)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18155

- Make a python decorator caffe2_flaky for caffe2 operator unit tests.
- The environment variable CAFFE2_RUN_FLAKY_TESTS are now used to mark flaky test mode

During test run,
- If flaky tests mode are on, only flaky tests are run
- If flaky tests mode are off, only non-flaky tests are run

Mark ctc_beam_search_decoder_op_test as flaky

Reviewed By: ezyang, salexspb

Differential Revision: D14468816

fbshipit-source-id: 579d0e8b5edcdce4ce18cdd5a1d92c2658849ef3
